### DISCUSSION TITLE
Harmonize Emacs indentation and LSP indentation

### DISCUSSION THREAD

**Discussion by wongandj** [joaotavora/eglot#1381] [General]
If you have a major-mode which sets its own formatting parameters (i.e. uses its own version of `tab-width`), then there is no way to communicate that to eglot so that requests to format the file, etc. use the proper formatting parameters.

### Reproduction Recipe

1. Ensure R and its `languageserver` package are installed along with the latest version of [ess](https://ess.r-project.org/).
2. Open a .R script and start the language server.
3. Use `M-x ess-set-style` or `C-c C-e s` to set the style to something with an `ess-indent-offset` of something other than `tab-width` (by default, this is 'RRR', with a 4-wide indent and call-aligned function arguments--very different from `tab-width`'s default 8).
4. Indent your code with `indent-region` or `indent-paragraph`*.
5. Use `M-x eglot-format-buffer` and watch it change all of the indentation again*.

*There is a [bug](https://github.com/REditorSupport/languageserver/issues/663#issue-2219442481) in `languageserver` right now in which the indentation linter is hardcoded to a 2-space indent, so it may be marked as incorrect indentation, even after the same language server (which the linter is a part of) just formatted it. But that is a `langageserver` problem.

### Solution Pointers

Looking over the source, it looks like maybe changing [these lines](https://github.com/joaotavora/eglot/blob/db91d58374627a195b731a61bead9b4f84a7e4bc/eglot.el#L3018-L3021) to refer to new package options which could be specified per-major-mode (e.g. in `eglot-server-programs`) would allow for users to specify which variables are the correct formatting parameters for each major mode. But, since I do not understand the codebase well, I am hesitant to write the patch myself for now.

**Comment by joaotavora** [comment]
There is no real harmony between Elisp `indent-region` and LSP "formatting".  There never was, and I doubt if it is even possible.  Elisp's indentation is about inserting or removing spaces or tabs at the beginning of the line, whereas LSP's is about adding/removing arbitrary whitespace (or maybe even more than that...).

The best you can hope for is baking an `indent-region-function`, but that didn't work very well last time I tried it: https://mail.gnu.org/archive/html/bug-gnu-emacs/2023-07/msg01541.html

**Reply by wongandj** [reply]
Yes, in my own investigation of the issue I've found that one of the issues preventing this is that LSPs are not required to make formatting decisions without context as `indent-region` does on RET, e.g. a code block or multiline function call might not get indented until it is closed. How do other editors do it? Or do they have "indentation wars" too?

**Reply by joaotavora** [reply]
> How do other editors do it? Or do they have "indentation wars" too?

🤷

**Comment by wongandj** [comment]
It still may be useful to expose those variables to bridge the gap a little. Even if it isn't perfect, a separate plist (listing major-modes and formatting parameters) could be made to keep things sanitary if it ever gets changed in the future. For now, I've written a small advice hack for init.el for myself and anyone else interested:
```
(advice-add 'eglot-format :around
	    (lambda (fn &rest r)
	      (let ((tab-width
		     (cond 
		      ((member major-mode '(ess-mode ess-r-mode)) ess-indent-offset))))
		(apply fn r))))
```
You can add more major modes by adding another clause to the `cond` call.

**Reply by joaotavora** [reply]
You can propose anything to bug-gnu-emacs@gnu.org, I won't veto it.  But I think this is just making a useless indirection.  Why not just set `tab-width` buffer-locally in your ers-mode thing?
